### ISSUE TITLE
Mountpoints and Other Dump Files broken after Upgrade

### ISSUE THREAD

**Issue by markus2330** [ElektraInitiative/libelektra#3575]
This is an obvious thing but just to have it also documented here but we also need to document it in the release notes!

We should recommend after the upgrade:

- [ ] fix the mountpoint configs, see below

Here are some errors people might get after the upgrade to Elektra containing keyname overhaul if they did not follow instructions above:

```
kdb ls user:/sw/elektra/dq/#0/lock
 Sorry, 171 warnings were issued ;(
        Sorry, module kdb issued the warning C01200:
        Installation: Could not create mountpoint with name 'system/hosts' and value system/hosts
        Sorry, module kdb issued the warning C01200:
        Installation: Backend has no mount point
        Sorry, module kdb issued the warning C01200:
        Installation: Could not create mountpoint with name 'system/info/elektra/constants' and value system/info/elektra/constants
        Sorry, module kdb issued the warning C01200:
        Installation: Backend has no mount point
        Sorry, module kdb issued the warning C01200:
        Installation: Could not create mountpoint with name 'system/info/elektra/contract/#0' and value system/info/elektra/contract/#0
        Sorry, module kdb issued the warning C01200:
        Installation: Dlopen failed. Could not load module libelektra-ini.so. Reason: libelektra-ini.so: cannot open shared object file: No such file or directory
        Sorry, module kdb issued the warning C01200:
        Installation: Could not load plugin ini in process plugin
        Sorry, module kdb issued the warning C01310:
        Internal: Could not reference back to plugin system:/elektra/plugins/ini
        Sorry, module kdb issued the warning C01200:
        Installation: Backend has no mount point
        Sorry, module kdb issued the warning C01200:
        Installation: Could not create mountpoint with name 'system/info/elektra/desktop' and value system/info/elektra/desktop
        Sorry, module The command kdb ls terminated unsuccessfully with the info:
key meta:/warnings/#_10/module was not found
Please report the issue at https://issues.libelektra.org
```

`kdb mount` outputs the mountpoints as before, without `:`. Here we see what is wrong:

```
kdb export system:/elektra/mountpoints
```

```
[user\\/sw\\/elektra\\/dq\\/#0/setplugins/#7#resolver]
 meta:/comment = A plugin
```

shows that actually everything is broken without the `:`.

If I want to umount something:
```
kdb umount system/hosts
#> The command kdb umount terminated unsuccessfully with the info:
#> std::bad_alloc
#> Please report the issue at https://issues.libelektra.org/
```

with backtrace:

```
#0  0x00007ffff7d84afd in __cxa_throw () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#1  0x000055555558a4e1 in kdb::Key::Key (this=0x7fffffffdf38, keyName=...) at /usr/include/c++/8/bits/exception.h:63
#2  0x00005555555855ff in Cmdline::createKey (this=this@entry=0x7fffffffe110, pos=pos@entry=0, allowCascading=allowCascading@entry=true) at /usr/include/c++/8/bits/basic_string.h:936
#3  0x0000555555612857 in UmountCommand::execute (this=0x55555567dd50, cl=...) at ./src/tools/kdb/umount.cpp:34
#4  0x000055555557c0bd in main (argc=<optimized out>, argv=0x7fffffffe488) at /usr/include/c++/8/bits/unique_ptr.h:342
```

So also umount of old-style mountpoints is not possible.

**Comment by mpranj** [comment]
Thank you for reporting!

Did you want to provide more support for the upgrade path? We could create a 0.9.4 release not based on master, but on the 0.9.3 tag and provide support for better upgrade path.

I'm not sure how many users this would affect though.

**Comment by markus2330** [comment]
No, I think we should keep the support at a minimum for now, as this is very time consuming, maybe with 1.0 release we make something. But as you say: only if someone cries and actually needs it :wink:

So for now: only some lines in the release notes.

**Comment by kodebach** [comment]
It is actually not terribly hard to fix almost all mountpoint configs **after the update**.

There are two places that contain the old syntax:
1. Every key below (and including) `system:/elektra/mountpoints/<MOUNTPOINT>` uses an old key names as `<MOUNTPOINT>`, if the mountpoint was created with `kdb mount`. Since this is just a name for the mountpoint, it is not a huge problem if this isn't updated. However, it is still a nice to have.
2. The value of all keys matching `system:/elektra/mountpoints/*/mountpoint` must be valid key names. This is what causes all the problems. But for almost all mountpoints the only issue is the missing colon `:`.

The following script will work for many configs, but I can't guarantee that it for absolutely all cases. I also can't guarantee that it won't break anything.

*You have been warned*

```sh
#! /usr/bin/env sh
kdb export system:/elektra/mountpoints ni > mountpoints.ini
kdb rm -r system:/elektra/mountpoints
sed -E 's~((^\[?|/mountpoint = )(user|system))((\\\\)?/)~\1:\4~g' export.ini > mountpoints_corrected.ini
kdb import system:/elektra/mountpoints ni < mountpoints_corrected.ini
```

The above script fixes both the `system:/elektra/mountpoints/*/mountpoint` values and the `system:/elektra/mountpoints/<MOUNTPOINT>` names. If you only want to fix the values and keep the old names, you can use this (simpler) `sed` expression instead:

*Still no guarantees*

```sed
s~(/mountpoint = (user|system))/~\1:/~g
```

We may want to create a separate pinned issue that just contains these instructions, so users/developers don't have to search for a solution.

**Comment by mpranj** [comment]
> It is actually not terribly hard to fix almost all mountpoint configs after the update.

If it can be fixed afterwards with a relatively simple script it would of course be ideal. Thank you for providing the solution!

**Comment by markus2330** [comment]
Thank you, very good input, so actually only one thing needs to be done before the upgrade (fixing the dump files) and one thing after the upgrade (fixing the absolute name in the mountpoints).

If we have time, we should write a proper script, that iterates over the KeySet and supports all namespaces. For me @kodebach script worked, I slightly updated it to make it a bit more safe. If we have nothing else till the release, the script is also okay as-is (with a disclaimer).

**Comment by kodebach** [comment]
> before the upgrade (fixing the dump files)

I don't think that is necessary... Dump can read the v1 format and will properly update, as long as all v1 format files use the old keyname syntax:

https://github.com/ElektraInitiative/libelektra/blob/7cba2f97152ce272307a14fe95232394b80b181d/src/plugins/dump/dump.cpp#L182-L185

 (see also `test_v1_*` in `testmod_dump.c`)

There might still be issues with keys that contain e.g. `/#10/`, but that isn't fixed by updating dump files to v2 before the keyname upgrade either. These kinds of bugs can only be fixed manually.

---

Also, I would recommend that at least during testing you enable the logger and check syslog/stderr. A warning will be logged, every time `key(V)New` failed because of a malformed key name:

https://github.com/ElektraInitiative/libelektra/blob/7cba2f97152ce272307a14fe95232394b80b181d/src/libs/elektra/key.c#L314-L319

Also setting, a breakpoint on `key.c:316` is a good starting point for debugging these keyname issues.

**Comment by markus2330** [comment]
> Dump can read the v1 format and will properly update

Yes, works for me. I updated the top-post.
